server/channels/store/sqlstore/oauth_outgoing_connection_store.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package sqlstore
+
+import (
+	"database/sql"
+
+	"github.com/mattermost/mattermost/server/public/model"
+	"github.com/mattermost/mattermost/server/public/shared/request"
+	"github.com/mattermost/mattermost/server/v8/channels/store"
+	"github.com/pkg/errors"
+)
+
+type SqlOAuthOutgoingConnectionStore struct {
+	*SqlStore
+}
+
+func NewSqlOAuthOutgoingConnectionStore(sqlStore *SqlStore) store.OAuthOutgoingConnectionStore {
+	return &SqlOAuthOutgoingConnectionStore{sqlStore}
+}
+
+func (s *SqlOAuthOutgoingConnectionStore) SaveConnection(c request.CTX, conn *model.OAuthOutgoingConnection) (*model.OAuthOutgoingConnection, error) {
+	if conn.Id != "" {
+		return nil, store.NewErrInvalidInput("OAuthOutgoingConnection", "Id", conn.Id)
+	}
+
+	conn.PreSave()
+	if err := conn.IsValid(); err != nil {
+		return nil, err
+	}
+
+	if _, err := s.GetMasterX().NamedExec(`INSERT INTO OAuthOutgoingConnection
+	(Id, Name, ClientId, ClientSecret, OAuthTokenURL, GrantType, Audiences)
+	VALUES
+	(:Id, :Name, :ClientId, :ClientSecret, :OAuthTokenURL, :GrantType, :Audiences)`, conn); err != nil {
+		return nil, errors.Wrap(err, "failed to save OAuthOutgoingConnection")
+	}
+	return conn, nil
+}
+
+func (s *SqlOAuthOutgoingConnectionStore) UpdateConnection(c request.CTX, conn *model.OAuthOutgoingConnection) (*model.OAuthOutgoingConnection, error) {
+	if conn.Id == "" {
+		return nil, store.NewErrInvalidInput("OAuthOutgoingConnection", "Id", conn.Id)
+	}
+
+	conn.PreUpdate()
+	if err := conn.IsValid(); err != nil {
+		return nil, err
+	}
+
+	if _, err := s.GetMasterX().NamedExec(`UPDATE OAuthOutgoingConnection SET
+	Name=:Name, ClientId=:ClientId, ClientSecret=:ClientSecret, OAuthTokenURL=:OAuthTokenURL, GrantType=:GrantType, Audiences=:Audiences
+	WHERE Id=:Id`, conn); err != nil {
+		return nil, errors.Wrap(err, "failed to update OAuthOutgoingConnection")
+	}
+	return conn, nil
+}
+
+func (s *SqlOAuthOutgoingConnectionStore) GetConnection(c request.CTX, id string) (*model.OAuthOutgoingConnection, error) {
+	conn := &model.OAuthOutgoingConnection{}
+	if err := s.GetReplicaX().Get(conn, `SELECT * FROM OAuthOutgoingConnection WHERE Id=?`, id); err != nil {
+		if err == sql.ErrNoRows {
+			return nil, store.NewErrNotFound("OAuthOutgoingConnection", id)
+		}
+		return nil, errors.Wrap(err, "failed to get OAuthOutgoingConnection")
+	}
+	return conn, nil
+}
+
+func (s *SqlOAuthOutgoingConnectionStore) GetConnections(c request.CTX, offset, limit int) ([]*model.OAuthOutgoingConnection, error) {
+	conns := []*model.OAuthOutgoingConnection{}
+	if err := s.GetReplicaX().Select(&conns, `SELECT * FROM OAuthOutgoingConnection LIMIT ? OFFSET ?`, limit, offset); err != nil {
+		return nil, errors.Wrap(err, "failed to get OAuthOutgoingConnections")
+	}
+	return conns, nil
+}
```

Comment:
This approach is not efficient. `LIMIT/OFFSET` will still load all the rows in DB memory.

What is instead suggested is to do a cursor-based pagination. Essentially you do `select * from table order by id limit <>` as the first query, and then send the id of the last row for the next query and do `select * from table where id > <> order by id limit <>`.

Reply:
Hey @agnivade, thanks for the feedback. I will change it accordingly. 

Just as a note, I used the OAuthApps store as template/inspiration where the `OFFSET/LIMIT` is being used (among other places). Do we have tickets to change/track changing those to this approach as well?

Reply:
Yes, our current code is not efficient. Fixing everything will take a lot of work obviously. It's unlikely to be prioritized soon. But for any new code, we should always try to use cursor-based pagination.

Reply:
I've just submitted the change. Please take a look and let me know if that's better.